src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Formatting/Passes/New/CSharpFormattingPass.CSharpDocumentGenerator.cs

```diff
@@ -0,0 +1,641 @@
+    /// reapply it so we end up with the C# indentation and Html indentation combined.
+    /// </para>
+    /// <para>
+    /// For more complete examples, the full test log for every formatting test includes the generated C# document.
+    /// </para>
+    /// <para>
+    /// A final important note about this class, whilst it is a SyntaxVisitor, it is not intended to be a general
+    /// purpose one, and things won't work as expected if the Visit method is called on arbitrary nodes. The visit
+    /// methods are implemented with the assumption they will only see a node if it is the first one on a line of
+    /// a Razor file.
+    /// </para>
+    /// </remarks>
+    private sealed class CSharpDocumentGenerator
+    {
+        public static CSharpFormattingDocument Generate(RazorCodeDocument codeDocument, RazorFormattingOptions options)
+        {
+            using var _ = StringBuilderPool.GetPooledObject(out var builder);
+            using var lineInfo = new PooledArrayBuilder<LineInfo>(capacity: codeDocument.Source.Text.Lines.Count);
+
+            var generator = new Generator(codeDocument, options);
+
+            generator.Generate(builder, ref lineInfo.AsRef());
+
+            var text = SourceText.From(builder.ToString());
+
+            return new(text, lineInfo.DrainToImmutable());
+        }
+
+        private static string GetAdditionalLineComment(SourceSpan originalSpan)
+        {
+            // IMPORTANT: The format here needs to match the parse method below
+            return $"// {originalSpan.AbsoluteIndex} {originalSpan.Length}";
+        }
+
+        public static (int start, int length) ParseAdditionalLineComment(string comment)
+        {
+            var span = comment.AsSpan();
+            var toParse = span.Slice(span.IndexOf(' ') + 1);
+            var space = toParse.IndexOf(' ');
+
+#if NET8_0_OR_GREATER
+            var start = int.Parse(toParse[..space]);
+            var length = int.Parse(toParse[(space + 1)..]);
+#else
+            var start = int.Parse(toParse.Slice(0, space).ToString());
+            var length = int.Parse(toParse.Slice(space + 1).ToString());
+#endif
+            return (start, length);
+        }
+
+        private sealed class Generator(RazorCodeDocument codeDocument, RazorFormattingOptions options) : SyntaxVisitor<LineInfo>
+        {
+            private readonly SourceText _sourceText = codeDocument.Source.Text;
+            private readonly RazorCodeDocument _codeDocument = codeDocument;
+            private readonly bool _insertSpaces = options.InsertSpaces;
+            private readonly int _tabSize = options.TabSize;
+
+            private TextLine _currentLine;
+            private int _currentFirstNonWhitespacePosition;
+
+            // These are set in GetCSharpDocumentContents so will never be observably null
+            private StringBuilder _builder = null!;
```

Comment:
Just noting that this can be passed into the constructor and the field can be readonly now.

Reply:
I'll be honest, I have yet to find any implementation of this class that I actually like fully. I did this change at one point, but it felt like if I was doing that, I may as well just pass the `lineInfo` array in too, and do all of the work in the constructor. Which also felt wrong :)

Reply:
Yeah, I think that'd be wrong. :smile: FWIW, I was only noting that `_builder` _could_ be passed to the constructor. I don't think it's necessary since it's harder to pass the `PooledArrayBuilder<LineInfo>` there. However, if you wanted to pass both to the constructor, you could just grab an `ImmutableArray<LineInfo>.Builder` from `ArrayBuilderPool` and forego the PooledArrayBuidler` savings.

Reply:
That's right. Alnd if things become a problem using the default pool, formatting could create and own its own pool.